### PR TITLE
System Table for Optimizer Rule Stats

### PR DESCRIPTION
**Pull request by phd3**
#4504

I think a single-page pagesource would be okay, since the total size (from `page.getSizeInBytes()`) comes out to be `~389K`. (mainly from the map block for percentiles)

### CODE REVIEW

**Comment by findepi** [comment]
Can we have a system view or jmx view, and pull this information from JMX ("single source of truth")?

**Comment by martint** [comment]
No, the JMX interface is not guaranteed to be stable. Also, it relies on the JMX connector being installed. This would be a system table provided by the core engine

**Comment by sopel39** [comment]
This is related to: https://github.com/prestosql/presto/issues/2578. https://github.com/prestosql/presto/issues/2578 would be per query, so I would say https://github.com/prestosql/presto/issues/2578 is preferred.

**Comment by phd3** [comment]
@findepi Apart from the benefits @martint suggested, I think the system table would handle data types for current/future columns better (eg. the percentiles as a map that can enable more complex queries without need for string manipulation).

@sopel39 thanks for linking this ticket, agree that query level info is more useful to directly relate between the query and rule application.

One advantage of having a system table would be to look at the rule usage, performance, failures over time on a workload. But if that's not worth adding this code, I can close this PR.

**Comment by martint** [comment]
> One advantage of having a system table would be to look at the rule usage, performance, failures over time on a workload. But if that's not worth adding this code, I can close this PR.

It's definitely worth adding :)

File: presto-main/src/main/java/io/prestosql/connector/system/RuleStatsSystemTable.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        this.ruleStatsRecorder = requireNonNull(ruleStatsRecorder, "ruleStatsRecorder is null");
+        requireNonNull(metadata, "metadata is null");
+
+        this.ruleStatsTable = tableMetadataBuilder(TABLE_NAME)
+                .column("rule_name", VARCHAR)
+                .column("invocations", BIGINT)
+                .column("matches", BIGINT)
+                .column("failures", BIGINT)
+                .column("average_time", DOUBLE)
+                .column("time_distribution_percentiles", new InternalTypeManager(metadata).getType(mapType(DOUBLE.getTypeSignature(), DOUBLE.getTypeSignature())))
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/connector/system/RuleStatsSystemTable.java:68]: Don't use InternalTypeManager. That's an implementation detail of the metadata system:

 ```suggestion
                .column("time_distribution_percentiles", metadata.getType(mapType(DOUBLE.getTypeSignature(), DOUBLE.getTypeSignature())))
```

File: presto-main/src/main/java/io/prestosql/connector/system/RuleStatsSystemTable.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            DOUBLE.writeDouble(blockBuilders.get("average_time"), stats.getTime().getAvg());
+
+            BlockBuilder mapWriter = blockBuilders.get("time_distribution_percentiles").beginBlockEntry();
+            for (Map.Entry<Double, Double> percentile : stats.getTime().getPercentiles().entrySet()) {
+                DOUBLE.writeDouble(mapWriter, percentile.getKey());
+                DOUBLE.writeDouble(mapWriter, percentile.getValue());
+            }
+            blockBuilders.get("time_distribution_percentiles").closeEntry();
+
+            VARCHAR.writeString(blockBuilders.get("time_unit"), stats.getTime().getUnit().name());
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/connector/system/RuleStatsSystemTable.java:111]: Let's remove this column. All measurements are in the same unit, so we don't need to clarify what the unit is per row.

**Comment by martint** [comment]
There's a test failure:

```
2020-08-10T16:53:22.2469071Z tests               | 2020-08-10 22:38:22 INFO: FAILURE     /    sql_tests.testcases.system.showTablesSystemRuntime (Groups: system) took 0.1 seconds
2020-08-10T16:53:22.2482337Z tests               | 2020-08-10 22:38:22 SEVERE: Failure cause:
2020-08-10T16:53:22.2483245Z tests               | java.lang.AssertionError: Expected row count to be <4>, but was <5>; rows=[[nodes], [optimizer_rule_stats], [queries], [tasks], [transactions]]
2020-08-10T16:53:22.2486734Z tests               | 	at io.prestosql.tempto.internal.convention.sql.SqlQueryConventionBasedTest.test(SqlQueryConventionBasedTest.java:102)
2020-08-10T16:53:22.2486985Z tests               | 	at io.prestosql.tempto.internal.convention.ConventionBasedTestProxyGenerator$ConventionBasedTestProxy.test(ConventionBasedTestProxyGenerator.java:123)
2020-08-10T16:53:22.2487200Z tests               | 	at io.prestosql.tempto.system.showTablesSystemRuntime(Unknown Source)
2020-08-10T16:53:22.2494936Z tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-08-10T16:53:22.2495177Z tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-08-10T16:53:22.2495415Z tests               | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-08-10T16:53:22.2495631Z tests               | 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2020-08-10T16:53:22.2499109Z tests               | 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-08-10T16:53:22.2499478Z tests               | 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
2020-08-10T16:53:22.2499688Z tests               | 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
2020-08-10T16:53:22.2499896Z tests               | 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
2020-08-10T16:53:22.2507128Z tests               | 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
2020-08-10T16:53:22.2507413Z tests               | 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
2020-08-10T16:53:22.2508076Z tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-08-10T16:53:22.2508388Z tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-08-10T16:53:22.2508543Z tests               | 	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by martint** [comment]
Another one:

```
2020-08-12T06:59:24.7683951Z tests               | 2020-08-12 12:44:24 INFO: FAILURE     /    sql_tests.testcases.system.selectInformationSchemaColumns (Groups: system) took 0.7 seconds
2020-08-12T06:59:24.7979411Z tests               | 2020-08-12 12:44:24 SEVERE: Failure cause:
2020-08-12T06:59:24.7979613Z tests               | java.lang.AssertionError: Not equal rows:
2020-08-12T06:59:24.7980164Z tests               | 71 - expected: system|runtime|queries|query_id|varchar|YES|null|null|
2020-08-12T06:59:24.7980546Z tests               | 71 - actual:   system|runtime|optimizer_rule_stats|rule_name|varchar|YES|null|null|
2020-08-12T06:59:24.7981223Z tests               | 72 - expected: system|runtime|queries|state|varchar|YES|null|null|
2020-08-12T06:59:24.7981783Z tests               | 72 - actual:   system|runtime|optimizer_rule_stats|invocations|bigint|YES|null|null|
2020-08-12T06:59:24.7982280Z tests               | 73 - expected: system|runtime|queries|user|varchar|YES|null|null|
2020-08-12T06:59:24.7983146Z tests               | 73 - actual:   system|runtime|optimizer_rule_stats|matches|bigint|YES|null|null|
2020-08-12T06:59:24.7983868Z tests               | 74 - expected: system|runtime|queries|source|varchar|YES|null|null|
2020-08-12T06:59:24.7984243Z tests               | 74 - actual:   system|runtime|optimizer_rule_stats|failures|bigint|YES|null|null|
2020-08-12T06:59:24.7984590Z tests               | 75 - expected: system|runtime|queries|query|varchar|YES|null|null|
2020-08-12T06:59:24.7984958Z tests               | 75 - actual:   system|runtime|optimizer_rule_stats|average_time|double|YES|null|null|
2020-08-12T06:59:24.7986103Z tests               | 76 - expected: system|runtime|queries|resource_group_id|array(varchar)|YES|null|null|
2020-08-12T06:59:24.7986598Z tests               | 76 - actual:   system|runtime|optimizer_rule_stats|time_distribution_percentiles|map(double, double)|YES|null|null|
2020-08-12T06:59:24.7986979Z tests               | 77 - expected: system|runtime|queries|queued_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.7987334Z tests               | 77 - actual:   system|runtime|queries|query_id|varchar|YES|null|null|
2020-08-12T06:59:24.7987696Z tests               | 78 - expected: system|runtime|queries|analysis_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.7988400Z tests               | 78 - actual:   system|runtime|queries|state|varchar|YES|null|null|
2020-08-12T06:59:24.7988797Z tests               | 79 - expected: system|runtime|queries|planning_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.7989428Z tests               | 79 - actual:   system|runtime|queries|user|varchar|YES|null|null|
2020-08-12T06:59:24.8075550Z tests               | 80 - expected: system|runtime|queries|created|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8076460Z tests               | 80 - actual:   system|runtime|queries|source|varchar|YES|null|null|
2020-08-12T06:59:24.8077010Z tests               | 81 - expected: system|runtime|queries|started|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8077505Z tests               | 81 - actual:   system|runtime|queries|query|varchar|YES|null|null|
2020-08-12T06:59:24.8078015Z tests               | 82 - expected: system|runtime|queries|last_heartbeat|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8078520Z tests               | 82 - actual:   system|runtime|queries|resource_group_id|array(varchar)|YES|null|null|
2020-08-12T06:59:24.8079013Z tests               | 83 - expected: system|runtime|queries|end|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8079495Z tests               | 83 - actual:   system|runtime|queries|queued_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.8079973Z tests               | 84 - expected: system|runtime|queries|error_type|varchar|YES|null|null|
2020-08-12T06:59:24.8080462Z tests               | 84 - actual:   system|runtime|queries|analysis_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.8081131Z tests               | 85 - expected: system|runtime|queries|error_code|varchar|YES|null|null|
2020-08-12T06:59:24.8081667Z tests               | 85 - actual:   system|runtime|queries|planning_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.8082255Z tests               | 86 - expected: system|runtime|tasks|node_id|varchar|YES|null|null|
2020-08-12T06:59:24.8082739Z tests               | 86 - actual:   system|runtime|queries|created|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8083213Z tests               | 87 - expected: system|runtime|tasks|task_id|varchar|YES|null|null|
2020-08-12T06:59:24.8083700Z tests               | 87 - actual:   system|runtime|queries|started|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8084160Z tests               | 88 - expected: system|runtime|tasks|stage_id|varchar|YES|null|null|
2020-08-12T06:59:24.8084646Z tests               | 88 - actual:   system|runtime|queries|last_heartbeat|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8085120Z tests               | 89 - expected: system|runtime|tasks|query_id|varchar|YES|null|null|
2020-08-12T06:59:24.8085604Z tests               | 89 - actual:   system|runtime|queries|end|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8086078Z tests               | 90 - expected: system|runtime|tasks|state|varchar|YES|null|null|
2020-08-12T06:59:24.8086563Z tests               | 90 - actual:   system|runtime|queries|error_type|varchar|YES|null|null|
2020-08-12T06:59:24.8087033Z tests               | 91 - expected: system|runtime|tasks|splits|bigint|YES|null|null|
2020-08-12T06:59:24.8087572Z tests               | 91 - actual:   system|runtime|queries|error_code|varchar|YES|null|null|
2020-08-12T06:59:24.8088056Z tests               | 92 - expected: system|runtime|tasks|queued_splits|bigint|YES|null|null|
2020-08-12T06:59:24.8088528Z tests               | 92 - actual:   system|runtime|tasks|node_id|varchar|YES|null|null|
2020-08-12T06:59:24.8089009Z tests               | 93 - expected: system|runtime|tasks|running_splits|bigint|YES|null|null|
2020-08-12T06:59:24.8089490Z tests               | 93 - actual:   system|runtime|tasks|task_id|varchar|YES|null|null|
2020-08-12T06:59:24.8089977Z tests               | 94 - expected: system|runtime|tasks|completed_splits|bigint|YES|null|null|
2020-08-12T06:59:24.8090687Z tests               | 94 - actual:   system|runtime|tasks|stage_id|varchar|YES|null|null|
2020-08-12T06:59:24.8091186Z tests               | 95 - expected: system|runtime|tasks|split_scheduled_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.8091661Z tests               | 95 - actual:   system|runtime|tasks|query_id|varchar|YES|null|null|
2020-08-12T06:59:24.8092588Z tests               | 96 - expected: system|runtime|tasks|split_cpu_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.8093091Z tests               | 96 - actual:   system|runtime|tasks|state|varchar|YES|null|null|
2020-08-12T06:59:24.8093587Z tests               | 97 - expected: system|runtime|tasks|split_blocked_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.8094065Z tests               | 97 - actual:   system|runtime|tasks|splits|bigint|YES|null|null|
2020-08-12T06:59:24.8094552Z tests               | 98 - expected: system|runtime|tasks|raw_input_bytes|bigint|YES|null|null|
2020-08-12T06:59:24.8095039Z tests               | 98 - actual:   system|runtime|tasks|queued_splits|bigint|YES|null|null|
2020-08-12T06:59:24.8095520Z tests               | 99 - expected: system|runtime|tasks|raw_input_rows|bigint|YES|null|null|
2020-08-12T06:59:24.8095988Z tests               | 99 - actual:   system|runtime|tasks|running_splits|bigint|YES|null|null|
2020-08-12T06:59:24.8096499Z tests               | 100 - expected: system|runtime|tasks|processed_input_bytes|bigint|YES|null|null|
2020-08-12T06:59:24.8096986Z tests               | 100 - actual:   system|runtime|tasks|completed_splits|bigint|YES|null|null|
2020-08-12T06:59:24.8097537Z tests               | 101 - expected: system|runtime|tasks|processed_input_rows|bigint|YES|null|null|
2020-08-12T06:59:24.8098023Z tests               | 101 - actual:   system|runtime|tasks|split_scheduled_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.8098652Z tests               | 102 - expected: system|runtime|tasks|output_bytes|bigint|YES|null|null|
2020-08-12T06:59:24.8099209Z tests               | 102 - actual:   system|runtime|tasks|split_cpu_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.8099693Z tests               | 103 - expected: system|runtime|tasks|output_rows|bigint|YES|null|null|
2020-08-12T06:59:24.8100480Z tests               | 103 - actual:   system|runtime|tasks|split_blocked_time_ms|bigint|YES|null|null|
2020-08-12T06:59:24.8107532Z tests               | 104 - expected: system|runtime|tasks|physical_input_bytes|bigint|YES|null|null|
2020-08-12T06:59:24.8113881Z tests               | 104 - actual:   system|runtime|tasks|raw_input_bytes|bigint|YES|null|null|
2020-08-12T06:59:24.8120018Z tests               | 105 - expected: system|runtime|tasks|physical_written_bytes|bigint|YES|null|null|
2020-08-12T06:59:24.8125383Z tests               | 105 - actual:   system|runtime|tasks|raw_input_rows|bigint|YES|null|null|
2020-08-12T06:59:24.8183401Z tests               | 106 - expected: system|runtime|tasks|created|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8186863Z tests               | 106 - actual:   system|runtime|tasks|processed_input_bytes|bigint|YES|null|null|
2020-08-12T06:59:24.8187371Z tests               | 107 - expected: system|runtime|tasks|start|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8187708Z tests               | 107 - actual:   system|runtime|tasks|processed_input_rows|bigint|YES|null|null|
2020-08-12T06:59:24.8188091Z tests               | 108 - expected: system|runtime|tasks|last_heartbeat|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8188456Z tests               | 108 - actual:   system|runtime|tasks|output_bytes|bigint|YES|null|null|
2020-08-12T06:59:24.8188828Z tests               | 109 - expected: system|runtime|tasks|end|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8189201Z tests               | 109 - actual:   system|runtime|tasks|output_rows|bigint|YES|null|null|
2020-08-12T06:59:24.8193952Z tests               | 110 - expected: system|runtime|transactions|transaction_id|varchar|YES|null|null|
2020-08-12T06:59:24.8194335Z tests               | 110 - actual:   system|runtime|tasks|physical_input_bytes|bigint|YES|null|null|
2020-08-12T06:59:24.8197552Z tests               | 111 - expected: system|runtime|transactions|isolation_level|varchar|YES|null|null|
2020-08-12T06:59:24.8198154Z tests               | 111 - actual:   system|runtime|tasks|physical_written_bytes|bigint|YES|null|null|
2020-08-12T06:59:24.8198525Z tests               | 112 - expected: system|runtime|transactions|read_only|boolean|YES|null|null|
2020-08-12T06:59:24.8199061Z tests               | 112 - actual:   system|runtime|tasks|created|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8201759Z tests               | 113 - expected: system|runtime|transactions|auto_commit_context|boolean|YES|null|null|
2020-08-12T06:59:24.8202264Z tests               | 113 - actual:   system|runtime|tasks|start|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8202633Z tests               | 114 - expected: system|runtime|transactions|create_time|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8203030Z tests               | 114 - actual:   system|runtime|tasks|last_heartbeat|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8203437Z tests               | 115 - expected: system|runtime|transactions|idle_time_secs|bigint|YES|null|null|
2020-08-12T06:59:24.8208278Z tests               | 115 - actual:   system|runtime|tasks|end|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8208709Z tests               | 116 - expected: system|runtime|transactions|written_catalog|varchar|YES|null|null|
2020-08-12T06:59:24.8209125Z tests               | 116 - actual:   system|runtime|transactions|transaction_id|varchar|YES|null|null|
2020-08-12T06:59:24.8209537Z tests               | 117 - expected: system|runtime|transactions|catalogs|array(varchar)|YES|null|null|
2020-08-12T06:59:24.8215952Z tests               | 117 - actual:   system|runtime|transactions|isolation_level|varchar|YES|null|null|
2020-08-12T06:59:24.8216485Z tests               | 118 - expected: system|runtime|optimizer_rule_stats|rule_name|varchar|YES|null|null|
2020-08-12T06:59:24.8216991Z tests               | 118 - actual:   system|runtime|transactions|read_only|boolean|YES|null|null|
2020-08-12T06:59:24.8217411Z tests               | 119 - expected: system|runtime|optimizer_rule_stats|invocations|bigint|YES|null|null|
2020-08-12T06:59:24.8222017Z tests               | 119 - actual:   system|runtime|transactions|auto_commit_context|boolean|YES|null|null|
2020-08-12T06:59:24.8222505Z tests               | 120 - expected: system|runtime|optimizer_rule_stats|matches|bigint|YES|null|null|
2020-08-12T06:59:24.8222921Z tests               | 120 - actual:   system|runtime|transactions|create_time|timestamp(3)|YES|null|null|
2020-08-12T06:59:24.8223334Z tests               | 121 - expected: system|runtime|optimizer_rule_stats|failures|bigint|YES|null|null|
2020-08-12T06:59:24.8223758Z tests               | 121 - actual:   system|runtime|transactions|idle_time_secs|bigint|YES|null|null|
2020-08-12T06:59:24.8228333Z tests               | 122 - expected: system|runtime|optimizer_rule_stats|average_time|double|YES|null|null|
2020-08-12T06:59:24.8228755Z tests               | 122 - actual:   system|runtime|transactions|written_catalog|varchar|YES|null|null|
2020-08-12T06:59:24.8229196Z tests               | 123 - expected: system|runtime|optimizer_rule_stats|time_distribution_percentiles|map(double, double)|YES|null|null|
2020-08-12T06:59:24.8229595Z tests               | 123 - actual:   system|runtime|transactions|catalogs|array(varchar)|YES|null|null|
2020-08-12T06:59:24.8233538Z tests               | 	at io.prestosql.tempto.internal.convention.sql.SqlQueryConventionBasedTest.test(SqlQueryConventionBasedTest.java:102)
2020-08-12T06:59:24.8233777Z tests               | 	at io.prestosql.tempto.internal.convention.ConventionBasedTestProxyGenerator$ConventionBasedTestProxy.test(ConventionBasedTestProxyGenerator.java:123)
2020-08-12T06:59:24.8233988Z tests               | 	at io.prestosql.tempto.system.selectInformationSchemaColumns(Unknown Source)
2020-08-12T06:59:24.8234187Z tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-08-12T06:59:24.8249199Z tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-08-12T06:59:24.8249434Z tests               | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-08-12T06:59:24.8249643Z tests               | 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2020-08-12T06:59:24.8249799Z tests               | 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-08-12T06:59:24.8253832Z tests               | 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
2020-08-12T06:59:24.8254154Z tests               | 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
2020-08-12T06:59:24.8254354Z tests               | 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
2020-08-12T06:59:24.8254567Z tests               | 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
2020-08-12T06:59:24.8258471Z tests               | 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
2020-08-12T06:59:24.8258671Z tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-08-12T06:59:24.8258869Z tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-08-12T06:59:24.8259052Z tests               | 	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by phd3** [comment]
my bad, I'll fix soon and re-request the review after tests pass.

**Pull request merged**